src/penn_chime/models.py

```diff
@@ -72,187 +77,181 @@ def __init__(self, p: Parameters):
-        is dependent on a given doubling_time in state.
-        """
-        # get the predicted number of hospitalized today
-        predicted_current_hospitalized = self.census_df.hospitalized.loc[p.n_days_since_first_hospitalized]
+def get_argmin_ds(census_df: pd.DataFrame, current_hospitalized: float) -> float:
+    losses_df = (census_df.hospitalized - current_hospitalized) ** 2.0
+    return losses_df.argmin()
 
-        # compare against actual / user inputted number
-        # we shall optimize squared distance
-        return (p.current_hospitalized - predicted_current_hospitalized) ** 2
 
-    def loss_ds(self, p: Parameters) -> float:
-        """loss with respect to days_since """
+def get_beta(
+    intrinsic_growth_rate: float,
+    gamma: float,
+    susceptible: float,
+    relative_contact_rate: float
+) -> float:
+    return (
+        (intrinsic_growth_rate + gamma)
+        / susceptible
+        * (1.0 - relative_contact_rate)
+    )
 
-        ## get the predicted number hospitalized today
-        pred_current_hospitalized = self.census_df['hospitalized'].loc[p.n_days_since_first_hospitalized]
 
-        ## compare against the actual (user inputed) number
-        ## squared difference is the loss to be optimized
-        return (p.current_hospitalized - pred_current_hospitalized)**2
+def get_growth_rate(doubling_time: Optional[float]) -> float:
+    """Calculates average daily growth rate from doubling time."""
+    if doubling_time is None or doubling_time == 0.0:
+        return 0.0
+    return (2.0 ** (1.0 / doubling_time) - 1.0)
 
-    def argmin_ds(self, p: Parameters) -> float:
-        losses_df = (self.census_df.hospitalized - p.current_hospitalized) ** 2
-        return losses_df.argmin()
 
-    @staticmethod
-    def _intrinsic_growth_rate(doubling_time: Optional[float]) -> float:
-        if doubling_time is not None:
-            return (2.0 ** (1.0 / doubling_time) - 1.0) if doubling_time > 0.0 else 0.0
-        return 0.0
+def get_loss(census_df: DataFrame, current_hospitalized: float, n_days_since: int) -> float:
+    """Squared error: predicted vs. actual current hospitalized."""
+    predicted = census_df.hospitalized.loc[n_days_since]
+    return (current_hospitalized - predicted) ** 2.0
 
-    @staticmethod
-    def _beta(
-            intrinsic_growth_rate: float,
-            gamma: float,
-            susceptible: float,
-            relative_contact_rate: float) -> float:
-        return (
-            (intrinsic_growth_rate + gamma)
-            / susceptible
-            * (1.0 - relative_contact_rate)
-        )
 
-###################
-##  MODEL FUNCS  ##
-###################
 def sir(
     s: float, i: float, r: float, beta: float, gamma: float, n: float
 ) -> Tuple[float, float, float]:
     """The SIR model, one time step."""
     s_n = (-beta * s * i) + s
     i_n = (beta * s * i - gamma * i) + i
     r_n = gamma * i + r
+    assert s_n >= 0.0
+    assert i_n >= 0.0
+    assert r_n >= 0.0
```

Comment:
can we go `s_n = max(-beta * s * i + s, 0)`, etcetera here?

Reply:
is there a valueadd of getting AssertionErrors in the case where the computation is less than zero?

Reply:
The max function was profiled to be slower than the if branches.

I'll remove the assertions --just seeing if the model actually went below zero for out default settings.

Reply:
Actually, why are we adjusting negative values? If the resulting df has negative values, isn't that a sure sign that the input data is bad? Shouldn't we check the df after it is created and warn the user?

Reply:
I think this adjustment hid the bug we had earlier, in fact.